### PR TITLE
[TF-TRT] Fix TopK Combined NMS

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
CC: @bixia1 

in tftrt, the tf.image.combined_non_max_suppression op is replaced by a trt engine using batchedNMSplugin, with the minimal reproducible code piece, there's something not aligned.

TRT nms plugin has argument "topK", and we will only do IOU operation on the topK box in each class, so we set topK to boxes count in each class, to fix mismatch between TF and TRT.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -5556,13 +5556,13 @@ Status ConvertCombinedNMS(OpConverterParams* params) {
   TFAttrs attrs(node_def);
   bool share_location = (boxes_dims.d[1] == 1);
   const bool pad_per_class = attrs.get<bool>("pad_per_class");
-  int top_k;
+  const int top_k = boxes_dims.d[0];
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5559]: I have a test case that shows the error, but it's a bit convoluted and complex to be packaged in a unittest. I'll try to see what I can do ;)

**Comment by DEKHTIARJonathan** [comment]
@bixia1 as requested here is a small script to reproduce the issue:

```python
import os
import time

os.environ["TF_CPP_MIN_LOG_LEVEL"]="3"

import numpy as np
import tensorflow.compat.v1 as tf

tf.disable_v2_behavior()

tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)


N = 2048
num_classes = 90

pre_nms_boxes = tf.placeholder(tf.float32, name='boxes', shape=[1, N, 1, 4])
pre_nms_scores = tf.placeholder(tf.float32, name='scores', shape=[1, N, num_classes])
max_detetion_points = N
max_boxes_to_draw = 30
iou_threshold = 0.1
min_score_thresh = 0.001

post_nms_boxes, post_nms_scores, post_nms_classes, post_nms_num_valid_boxes = \
    tf.image.combined_non_max_suppression(
        pre_nms_boxes,
        pre_nms_scores,
        max_output_size_per_class=max_detetion_points,
        max_total_size=max_boxes_to_draw,
        iou_threshold=iou_threshold,
        score_threshold=min_score_thresh,
        pad_per_class=False
    )

np.random.seed(666)
np_boxes = np.random.rand(1, 2048, 1, 4)
np_scores = np.random.rand(1, 2048, 90)

with tf.Session() as sess:
    tf_output = sess.run([
            post_nms_boxes,
            post_nms_scores,
            post_nms_classes,
            post_nms_num_valid_boxes
        ],
        feed_dict={pre_nms_boxes: np_boxes, pre_nms_scores: np_scores}
    )

with tf.Session() as sess:
    output = [
        tf.identity(i)
        for i in [
            post_nms_boxes,
            post_nms_scores,
            post_nms_classes,
            post_nms_num_valid_boxes
        ]
    ]

    output_node_names = [node.name.split(':')[0] for node in output]

    graphdef = tf.graph_util.convert_variables_to_constants(
        sess, sess.graph_def, output_node_names
    )

    fetches = [pre_nms_boxes.name, pre_nms_scores.name] + [i.name for i in output]

    from tensorflow.python.compiler.tensorrt import trt_convert as trt

    converter = trt.TrtGraphConverter(
        nodes_blacklist=[t.split(':')[0] for t in fetches],
        input_graph_def=graphdef,
        max_workspace_size_bytes=1 << 32,    # 4Gb
        minimum_segment_size=1,
        precision_mode='FP32')
    infer_graph = converter.convert()

    goutput = tf.import_graph_def(infer_graph, return_elements=fetches)

    inputs, output = goutput[:2], goutput[2:]

    trt_output = sess.run(
        output,
        feed_dict={inputs[0]: np_boxes, inputs[1]: np_scores}
    )


print("tf_output  - post_nms_classes:", tf_output[2])
print("trt_output - post_nms_classes:", trt_output[2])
```

If you run with current upstream tensorflow:
```bash
$ python main.py 

WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/compat/v2_compat.py:88: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
tf_output - post_nms_classes: [[89.  6. 85. 51. 68.  1.  1. 78. 51. 27. 61. 63. 30. 49. 78. 45.  8. 46.
  76. 10. 35. 37. 24. 77. 13. 16. 35. 27. 28. 25.]]
trt_output - post_nms_classes: [[89.  6. 85. 51. 68.  1.  1.  1. 78. 51. 27. 85. 61. 63. 30. 49. 78. 45.
   8. 46. 76. 10. 35. 37. 24. 77. 13. 16. 35. 27.]]
```

As you can see the two results are different.

```bash
$ python main.py 

WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
tf_output  - post_nms_classes: [[89.  6. 85. 51. 68.  1.  1. 78. 51. 27. 61. 63. 30. 49. 78. 45.  8. 46.
  76. 10. 35. 37. 24. 77. 13. 16. 35. 27. 28. 25.]]
trt_output - post_nms_classes: [[89.  6. 85. 51. 68.  1.  1. 78. 51. 27. 61. 63. 30. 49. 78. 45.  8. 46.
  76. 10. 35. 37. 24. 77. 13. 16. 35. 27. 28. 25.]]
```

And now with this change we are good ;)

**Comment by bixia1** [comment]
Thanks for the test case. Can we find a way to turn in into a test in the tree?

**Comment by gbaned** [comment]
@DEKHTIARJonathan  Can you please check @bixia1's comments and keep us posted. Thanks!

**Review by bixia1 — DISMISSED** [review]

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please address Ubuntu Sanity errors? Thanks!

**Comment by DEKHTIARJonathan** [comment]
@gbaned styling fixed ;)

**Review by bixia1 — DISMISSED** [review]

**Comment by tfeher** [comment]
@googlebot I consent.

**Comment by tfeher** [comment]
I have improved the converter and updated the C++ unit test of CombinedNMS converter. The changes are in branch https://github.com/tfeher/tensorflow/tree/tftrt/topk_combined_nms_fix

@DEKHTIARJonathan  I did not want to force push to your branch, but it would make sense to continue the discussion here. Please consider pulling the changes from my branch here. (Note @bixia1 that my branch contains a changeset of #39990, that should be reviewed before this PR).

Here are the changes:
- We have to set top_k so that keep_top_k <= top_k, otherwise the plugin does not work with TRT 7.0
- Enabled explicit batch mode for the converter, added test for it. Dynamic shapes are not supported by the plugin.
- Added test to check the clib_boxes attribute usage
- Added test with nonzero score threshold
- Added test where the boxes are defined as y_max, x_max, y_min, x_min. This test will fail with TRT 7.1.2 or earlier and highlights that the CombinedNMS plugin is not compatible with the TF definition of the operator in these versions of TRT. @DEKHTIARJonathan  please check if this test works with 7.1.3.
  Because of this error, it is not safe to convert CombinedNMS before 7.1.3. We should probably add an
  ```c++
  #if !IS_TRT_VERSION_GE(7, 1, 3, 0)
    errors::Unimplemented("TensorRT BatchedNMS Plugin reqires TensorRT 7.1.3 or later");
  #endif
  ```
to the converter.  What is your opinion @bixia1?

**Comment by DEKHTIARJonathan** [comment]
@tfeher please feel free to force push to my branch ;) I will test with TRT 7.1.3

**Comment by tfeher** [comment]
Updated this PR with the changes mentioned [above](https://github.com/tensorflow/tensorflow/pull/40062#issuecomment-657263292).

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -5698,23 +5742,33 @@ Status ConvertCombinedNMS(OpConverterParams* params) {
...
+  // TRT op is_normalized=False treats input corrdinates as pixels and
+  // calculates width/height as (max - min + 1).
+  //
+  // TF op CombinedNonMaxSuppression doesn't care about the normalization and
+  // calculates width/height  as (max-min).
+  //
+  // We set is_normalized = true to be consistent with TF IOU calculaton.
   const bool is_normalized = true;
+
   // Set plugin fields and the field collection
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5754]: This comment line here doesn't look very helpful to me now. We can either move it closer to "fields" or remove it.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 3]
```diff
@@ -1606,28 +1606,35 @@ class OpConverterTest : public ::testing::Test {
   }
 
   // Helper method to run both validation and conversion, and check the output
-  // shape.
-  void RunValidationAndConversion(const NodeDef& node_def, const Status& status,
-                                  const char* output_name,
-                                  const std::vector<int>& exp_out_dims) {
+  // shape. TODO(tfeher have variant that can check multiple outputs
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1609]: Let's fix the format for TODO and the description:
TODO(tfeher): Add a variant to check multiple outputs.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 4]
```diff
@@ -1606,28 +1606,35 @@ class OpConverterTest : public ::testing::Test {
...
+          string name = i == 0 ? output_name : StrCat(output_name, ":", i);
+          TF_EXPECT_OK(GetTensorOrWeights(name.c_str(), &output));
+          ASSERT_TRUE(output.is_tensor());
+          if (!exp_out_dims[i].empty()) {
+            // We only check output shape implicit batch mode. In dynamic shape
+            // mode we need to wait for the concrate input shapes to be defined
+            // (by setBindingDimensions before enqueue) before we can check
+            // whether the output dims are equal.
+            //
+            // TODO(tamas) enable this check in explicit_batch_mode
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1628]: Similar to the above, please fix the format for TODO.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 2 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: I have removed it.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 3 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: Actually this TODO is out of date, the modified code already checks multiple outputs. I have removed it.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 4 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: I have fixed the format.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia for the review, I have improved the comments in the C++ part. @DEKHTIARJonathan did you have a chance to check conversion with TRT 7.1.3?

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@DEKHTIARJonathan  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by bixia1** [comment]
Yes, I think @tfeher is working on it.

**Comment by tfeher** [comment]
I have updated the PR, from my side it is ready @bixia1.
- Checked that it works with TRT 7.1.3.
- Disabled conversion with any earlier TRT version. TRT plugin is inconsitent with TF in earlier versions and  [this](https://github.com/DEKHTIARJonathan/tensorflow/blob/269bdf6756d4cd5ca175c4cf85583fd85e40f485/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L3131-L3134) test would fail. FYI @DEKHTIARJonathan

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@DEKHTIARJonathan Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@DEKHTIARJonathan Any update on this PR? Please. Thanks!

**Comment by DEKHTIARJonathan** [comment]
Do not close, WIP

**Comment by DEKHTIARJonathan** [comment]
Steps to confirm:
- [ ] Checked that it works with TRT 7.1.3.
- [ ] Disabled conversion with any earlier TRT version. TRT plugin is inconsitent with TF in earlier versions and this test would fail. FYI @DEKHTIARJonathan

**Comment by DEKHTIARJonathan** [comment]
@bixia1 I have finished testing and doing the last modifications. We are good to go.
If fine with you we can merge ;)

**Comment by DEKHTIARJonathan** [comment]
@bixia1 @tfeher we are good to merge ;)

[combined_nms_test.log](https://github.com/tensorflow/tensorflow/files/5460526/combined_nms_test.log)

Attached is the log of the unittest. All passed and good ;)

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — CHANGES_REQUESTED** [review]
The added tests fail with TRT 5 with error message:
Value of: status.error_message()
Expected: has substring "TensorRT BatchedNMS Plugin reqires TensorRT 7.1.3 or later"
  Actual: "Op type CombinedNonMaxSuppression is not supported."
Unimplemented: Op type CombinedNonMaxSuppression is not supported.
Google Test trace:
third_party/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:3202: Test 1: Original test
Stack trace:
0x56016fcd650d: tensorflow::tensorrt::convert::ExpectStatus(tensorflow::Status, tensorflow::error::Code, char const*) @ ../sysdeps/x86_64/start.S:121
0x56016fcec3b7: tensorflow::tensorrt::convert::OpConverterTest::RunValidationAndConversion(tensorflow::NodeDef const&, tensorflow::error::Code, char const*, bool) @ ??:??
0x56016fd7a5b6: tensorflow::tensorrt::convert::OpConverterTest::RunValidationAndConversion(tensorflow::NodeDef const&, tensorflow::Status const&, char const*, std::__u::vector<std::__u::vector<int, std::__u::allocator<int> >, std::__u::allocator<std::__u::vector<int, std::__u::allocator<int> > > > const&) @ ??:??
0x56016fd1849c: tensorflow::tensorrt::convert::OpConverterTest1_ConvertCombinedNMS_Test::TestBody() @ ??:??
0x56017038b981: testing::Test::Run() @ ??:??
0x56017038cc28: testing::TestInfo::Run() @ ??:??
... Google Test internal frames ...

**Comment by bixia1** [comment]
One more test failure, tensorflow/python/compiler/tensorrt:combined_nms_test with TRT 7.1.3:

Traceback (most recent call last):
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/framework/test_util.py", line 1397, in decorated
    return f(self, *args, **kwargs)
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1107, in _Test
    self.RunTest(run_params)
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1030, in RunTest
    GraphState.INFERENCE)
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 910, in _VerifyGraphDef
    graph_state)
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 848, in _VerifyGraphDefV1
    self._VerifyConnections(expected_engines, original_gdef, gdef_to_verify)
  File "/build/work/12b5ec9ab1fdb4841ff23eda896242132066/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 670, in _VerifyConnections
    (sorted(expected_input_map.items()), sorted(actual_input_map.items())))
AssertionError: {'TRTEngineOp_0': {'pre_nms_scores', 'pre_nms_boxes'}, 'input_0': set(), 'input_1': set(), 'output_0': {'TRTEngineOp_0'}, 'output_1': {'TRTEngineOp_0'}, 'output_2': {'TRTEngineOp_0'}, 'output_3': {'TRTEngineOp_0'}, 'pre_nms_boxes': {'input_0'}, 'pre_nms_scores': {'input_1'}} != {'TRTEngineOp_0': {'input_1', 'input_0'}, 'input_0': set(), 'input_1': set(), 'output_0': {'TRTEngineOp_0'}, 'output_1': {'TRTEngineOp_0'}, 'output_2': {'TRTEngineOp_0'}, 'output_3': {'TRTEngineOp_0'}} (
expected:
[('TRTEngineOp_0', {'pre_nms_scores', 'pre_nms_boxes'}), ('input_0', set()), ('input_1', set()), ('output_0', {'TRTEngineOp_0'}), ('output_1', {'TRTEngineOp_0'}), ('output_2', {'TRTEngineOp_0'}), ('output_3', {'TRTEngineOp_0'}), ('pre_nms_boxes', {'input_0'}), ('pre_nms_scores', {'input_1'})]
vs actual:
[('TRTEngineOp_0', {'input_1', 'input_0'}), ('input_0', set()), ('input_1', set()), ('output_0', {'TRTEngineOp_0'}), ('output_1', {'TRTEngineOp_0'}), ('output_2', {'TRTEngineOp_0'}), ('output_3', {'TRTEngineOp_0'})])
repr() of differing entries:
'TRTEngineOp_0': {'pre_nms_scores', 'pre_nms_boxes'} != {'input_1', 'input_0'}

Missing entries:
'pre_nms_boxes': {'input_0'}
'pre_nms_scores': {'input_1'}

**Comment by gbaned** [comment]
@DEKHTIARJonathan  Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by DEKHTIARJonathan** [comment]
@bixia1 the C++ unittest wasn't properly deactivated with TRT < 7.1.3.

Fixed has been applied and commits squashed: https://github.com/tensorflow/tensorflow/pull/40062/files#diff-
b9ae60f058d9ae72ddc0625fa93ff1b8441e57db5695e8444d18becdb3abea2dL2976-R3002

Please run the CI again ;) And we can merge if good

**Review by bixia1 — APPROVED** [review]

**Comment by bixia1** [comment]
It seems to me that this failure is still reproducible
https://github.com/tensorflow/tensorflow/pull/40062#issuecomment-722816879

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by DEKHTIARJonathan** [comment]
Reassigning this issue to @tfeher

**Comment by DEKHTIARJonathan** [comment]
@tfeher could we finalized the last few bits of that one ?

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tfeher** [comment]
I have updated the PR. I could not reproduce the failing test yet. According to the log it is caused by the V1 graph def verification (`_VerifyGraphDefV1`) which is known to be finicky. The nodes in the converted graphs has to match the lists given by `ExpectedEnginesToBuild`. I have removed the extra identity nodes added by the new test, in principle this should fix the error.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**